respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,316 @@
+        raise ValueError(
+            "Replacement functions can only be matched 1:1 or 1:n with sets of "
+            "empirical moments."
+        )
+
+    elif len(replace_nans) > len(empirical_moments):
+        raise ValueError(
+            "There are more replacement functions than sets of empirical moments."
+        )
+
+    else:
+        pass
+
+    if len(calc_moments) != len(empirical_moments):
+        raise ValueError(
+            "Number of functions to calculate simulated moments must be equal to "
+            "the number of sets of empirical moments."
+        )
+
+    msm_func = functools.partial(
+        msm,
+        simulate=simulate,
+        calc_moments=calc_moments,
+        replace_nans=replace_nans,
+        empirical_moments=empirical_moments,
+        weighting_matrix=weighting_matrix,
+        return_scalar=return_scalar,
+    )
+
+    return msm_func
+
+
+def msm(
+    params,
+    simulate,
+    calc_moments,
+    replace_nans,
+    empirical_moments,
+    weighting_matrix,
+    return_scalar,
+):
+    """Loss function for msm estimation.
+
+    Parameters
+    ----------
+    params : pandas.DataFrame or pandas.Series
+        Contains model parameters.
+    simulate : callable
+        Function used to simulate data for msm estimation.
+    calc_moments : list
+        List of function(s) used to calculate simulated moments. Must match length of
+        empirical_moments i.e. calc_moments contains a moments function for each item in
+        empirical_moments.
+    replace_nans : list
+        List of functions(s) specifying how to handle missings in simulated_moments.
+        Must match length of empirical_moments.
+    empirical_moments : list
+        Contains the empirical moments calculated for the observed data. Each item in
+        the list constitutes a set of moments saved to a pandas.DataFrame or
+        pandas.Series. Index of pandas.DataFrames can be of type MultiIndex, but columns
+        cannot.
+    weighting_matrix : numpy.ndarray
+        Square matrix of dimension (NxN) with N denoting the number of
+        empirical_moments. Used to weight squared moment errors.
+    return_scalar : bool
+        Indicates whether to return moment error vector (False) or weighted square
+        product of moment error vector (True).
+
+    Returns
+    -------
+    out : pandas.Series or float
+        Scalar or moment error vector depending on value of return_scalar.
+
+    """
+    empirical_moments = copy.deepcopy(empirical_moments)
+
+    df = simulate(params)
+    simulated_moments = [func(df) for func in calc_moments]
+
+    simulated_moments = [
+        sim_mom.reindex_like(emp_mom)
+        for emp_mom, sim_mom in zip(empirical_moments, simulated_moments)
+    ]
+
+    simulated_moments = [
+        func(mom) for mom, func in zip(simulated_moments, replace_nans)
+    ]
+
+    flat_empirical_moments = _flatten_index(empirical_moments)
```

Comment:
Agreed, I have thought this as well. Flattening the empirical moments in msm when that really only needs to happen once probably takes up unnecessary time over multiple iterations in an optimizer. How exactly would you aggregate calc_moments and replace_nans? I think we could probably change the order of arguments within msm to 'simulate > calculate moments > flatten > reindex > replace'. This would allow us to move the flattening of the empirical moments to get_msm_func and eliminate the list comprehension for reindexing. However, I am not sure on how to handle the replacement step then since we would have to apply the replacement functions to different parts of the flattend moments vector?

Reply:
I am just guessing, but I think all operations here are completely irrelevant in terms of performance. The operations are not complex and might take a higher percentage of the total runtime for smaller models, but smaller models will be fast anyway. For larger models the solution dominates everything.

It should work to flatten the empirical moments beforehand and reindex with the flattened index.

But, I would not do the replacement on the flattened index. Then, the user needs to write a replacement function which works on a subset of an index. At best, the user does not need to know about the flattened index. Instead, if you have separated series' and dataframes in ``simulated_moments``, replacement functions can be as simple as ``lambda x: x.fillna(0)``.

Reply:
Ok but that implies that we have to stick to the currennt implementation since we have to apply replace nans  after we perform the reindexing step? 
I get your arguments and am also fine with the current solution!